File: AlgoExpertPy/Arrays/8-Move Element To End.py
def moveElementToEnd(array, toMove):
    leftIdx = 0
    rightIdx = len(array) - 1
    
    print(array)
    while leftIdx < rightIdx:
        print(leftIdx, array[leftIdx], rightIdx, array[rightIdx])
        if array[leftIdx] == toMove and array[rightIdx] != toMove:
            array[leftIdx], array[rightIdx] = array[rightIdx], array[leftIdx]
        elif array[leftIdx] == toMove and array[rightIdx] == toMove:
            rightIdx -= 1
            continue
        elif array[leftIdx] != toMove and array[rightIdx] != toMove:
            leftIdx += 1
            continue
        leftIdx += 1
        rightIdx -= 1
    return array

def moveElementToEnd2(array, toMove):
    leftIdx = 0
    rightIdx = len(array) - 1

    while leftIdx < rightIdx:
        if array[rightIdx] == toMove:
            rightIdx -= 1
            continue
        if array[leftIdx] == toMove:
            array[leftIdx], array[rightIdx] = array[rightIdx], array[leftIdx]
        leftIdx += 1
    return array


print(moveElementToEnd2([2, 1, 2, 6, 8, 9, 10, 2], 2))
    

File: AlgoExpertJS/8-Move Element To End.js
function moveElementToEnd(array, toMove) {
    var leftIdx = 0
    var rightIdx = array.length - 1
    while (leftIdx <= rightIdx) {
        if (array[leftIdx] == toMove) {
            if (array[rightIdx] == toMove) {
                rightIdx--
            } else {
                [array[leftIdx], array[rightIdx]] = [array[rightIdx], array[leftIdx]]
            }
        } else {
            leftIdx++
        }
    }
    return array
}
var arr = [3, 3, 2, 2, 2, 5, 4, 3, -5, -2, 1, 2, 2, 2, 2, 2, 2, 2, 3, 4, 5, 6, 7, 8, 9, 10]
moveElementToEnd(arr, 3)
console.log(arr)